### PR TITLE
JSON formatting (part 2) [RDY]

### PR DESCRIPTION
**Pull request by mugling**
### Summary

Continues from #17431
- Extends support to include many more types and adds further unit tests in `examples`
- 1500% faster via caching of already matched contexts (b634607)
- Supports `ARRAY` flag to support canonical output (see below).
### Canonical arrays

Currently some nodes can expressed in multiple equivalent forms:

```
"flags": "NO_RELOAD"
"flags": [ "NO_RELOAD" ]
```

If we ever want to support automatic generation of JSON content (eg. #1004) this will pose a problem as a different form could be serialized out than the form originally read despite the contents being otherwise unchanged. This would result in large diffs and precludes the development or usage of such tooling.

The `ARRAY` flag provides a solution by converting the first form to the second and is applied initially for item `flags` only.

In keeping with [Postels law](http://en.wikipedia.org/wiki/Robustness_principle) the game binary accepts both forms but our automated tooling (including this linter) should output only one canonical form. Via this approach automated tools don't cycle between alternative forms, the existing JSON data remains usable without changes and mod authors external to the project need take no action.
### Further development

This new ruleset identifies a number of problems with the game JSON data. These will be pushed separately as this PR includes only an updated ruleset and enhancements to the linter.

The plan is to submit a series of PR's extending the ruleset and providing `examples` of it's output until we have complete coverage. At this point we can begin using it to lint particularly ugly sections of JSON (especially some of the mods).

Given the improvements in performance it is now reasonably practical to use `format.pl -q [file]` to check a file is both valid JSON and contains no unused fields. Once we have 100% coverage I'd like to extend the `Makefile` to include this check for all game JSON data. This wouldn't deal with any formatting issues but is a big step towards detecting future typos.

### CODE REVIEW

**Comment by mugling** [comment]
To test check the the unit tests run and that `examples` are satisfactory

**Comment by Coolthulhu** [comment]
> If we ever want to support automatic generation of JSON content (eg. #1004) this will pose a problem as a different form could be serialized out than the form originally read despite the contents being otherwise unchanged. This would result in large diffs and precludes the development or usage of such tooling.

Would it be possible to automatically cancel further changes after `n` changed lines?

**Comment by mugling** [comment]
> Would it be possible to automatically cancel further changes after n changed lines?

Can you rephrase?

**Comment by Coolthulhu** [comment]
* Initialize some upper limit on total diff size
* Scan json object
* Parse it
* Create a diff with the original json object
* If the diff has less than `total_diff_size - current_diff_size`, add it to output file and increase `current_diff_size` by size of this diff
* Otherwise add the original object to the output and only print warnings about incomplete parsing

That way we could submit partial styling of each file instead of having to deal with 1000+ line diffs.

**Comment by mugling** [comment]
@Coolthulhu no, but once I've written the ruleset this would be a reasonable next step

**Comment by mugling** [comment]
Have further extended ruleset which is now able to parse all of generic guns. Ready for merge

**Comment by mugling** [comment]
Initial cleanup required for blazemod:
```
sed -i s/"comment"/"\/\/"/ data/mods/blazemod/*
sed -i s/"comment"/"\/\/"/ data/mods/blazemod/*
sed -i s/"mod_type"/"mod-type"/ data/mods/blazemod/*
sed -i s/"skill_sec"/"skills_required"/ data/mods/blazemod/*
sed -i /"wheel_width"/d data/mods/blazemod/*
```
Then drop the `ammo` field from `blob_tank` in `blaze_blob.json`

Finally:
```
./tools/cleanup.sh none data/mods/blazemod/*
```

**Comment by Coolthulhu** [comment]
I'm invoking `format.pl -q [not OK file here]` and it doesn't print any complaints.

**Comment by mugling** [comment]
> I'm invoking format.pl -q [not OK file here] and it doesn't print any complaints.

what is the exact command your running. I get

For non-existent file `badfoo` I get:
```
Can't open badfoo: No such file or directory at ./tools/format/format.pl line 102.
```

For `notjson.txt`: 
```
ERROR: Syntax error on line 1
```

**Comment by Coolthulhu** [comment]
OK, the file could actually be OK. Other file prints error properly.
